site/tests/legacy/test_deposit_api.py

```diff
@@ -0,0 +1,316 @@
+            # See https://github.com/zenodo/zenodo-rdm/pull/289
+        ],
+        locations=[
+            dict(
+                place="London",
+                description="A nice location",
+                lat=10.2,
+                lon=5,
+            ),
+            dict(place="Lisbon", lat=5.1231221, lon=1.23232323),
+        ],
+        title="Test title",
+        upload_type="publication",
+    )
+
+
+@pytest.fixture
+def get_json():
+    """Function for extracting json from response."""
+
+    def inner(response, code=None):
+        """Decode JSON from response."""
+        data = response.get_data(as_text=True)
+        if code is not None:
+            assert response.status_code == code, data
+        return json.loads(data)
+
+    return inner
+
+
+@pytest.fixture
+def deposit_url():
+    """Deposit API URL."""
+    return f"/api{LegacyRecordResourceConfig.url_prefix}"
+
+
+def test_invalid_create(
+    test_app, client_with_login, es, deposit_url, get_json, headers, db
+):
+    """Test invalid deposit creation.
+
+    Test ported from https://github.com/zenodo/zenodo/blob/master/tests/unit/deposit/test_api_metadata.py.
+    """
+    client = client_with_login
+
+    # Invalid deposits.
+    cases = [
+        dict(unknownkey="data", metadata={}),
+        dict(metadata={}),
+    ]
+
+    for case in cases:
+        res = client.post(deposit_url, data=json.dumps(case), headers=headers)
+        assert res.status_code == 400, case
+
+    # No deposits were created
+    res = client.get(deposit_url, headers=headers)
+    assert res.json["hits"]["total"] == 0
+
+
+def test_input_output(
+    client_with_login,
+    es,
+    db,
+    headers,
+    deposit_url,
+    get_json,
+    test_data,
+    expected_record_metadata,
+):
+    """Rough validation of input against output data.
+
+    Test ported from https://github.com/zenodo/zenodo/blob/master/tests/unit/deposit/test_api_metadata.py.
+    """
+    client = client_with_login
+
+    # Create
+    res = client.post(deposit_url, data=json.dumps(test_data), headers=headers)
+    links = get_json(res, code=201)["links"]
+
+    # Get serialization.
+    data = get_json(client.get(links["self"], headers=headers), code=200)
+
+    # - fix known differences.
+    # DOI and recid have 2 as control number, since Concept DOI/recid are
+    # registered first
+    recid = data["id"]
+    expected_record_metadata.update(
+        {
+            "prereserve_doi": {"doi": f"10.5281/zenodo.{recid}", "recid": recid},
+        }
+    )
+
+    ignored_keys = set()
+
+    # doi is returned as a top level key (and not inside metadata)
+    assert data["doi"] == test_data["metadata"]["doi"]
+    ignored_keys.add("doi")
+
+    differences = list(
+        dictdiffer.diff(data["metadata"], expected_record_metadata, ignore=ignored_keys)
+    )
+
+    # Known issues:
+    # - related_identifiers relation expects 'isCitedBy' but returns lower cased.
```

Comment:
@slint these are the two failing scenarios:

- `isCitedBy` can be mapped back (or simply ignored)
- I think `additional_descriptions` might be failing due to a vocabulary missing, however this can easily be tested by running a local instance and executing the same request.